File: strategy/marvin/agent/defender/piggy.lua
local Base = require "agent/base/behavior"
local Piggy = Class("Agent.Defender.Piggy", Base)

local debug = require "../base/debug"

-- TODO use interceptPass as soon as it works
--local InterceptPass = require "task/defender/interceptpass"

local PiggyTask = require "task/defender/piggy"


function Piggy:_stop()
	self._opp = nil
end

function Piggy:check()
	local role = self._inbox.roleAssignment().trainer
	return role and role.name == "Piggy"
end

function Piggy:_updateTask()
	local newOpp = self._inbox.roleAssignment().trainer.params[1]
	local restartTask = newOpp ~= self._opp
	self._opp = newOpp

	debug.set("target", self._opp.id)

	return PiggyTask, { self._opp }, restartTask
end

return Piggy

File: strategy/typescript/glados/agent/defender/piggy.ts
/**************************************************************************
*   Copyright 2026 Robotics Erlangen e.V., Tobias Heineken                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
**************************************************************************/

import * as debug from "base/debug";
import { Robot } from "base/robot";

import { Behavior, TaskAssignment } from "glados/agent/base/behavior";
import { MessageType } from "glados/control/messaging";
import { Piggy as PiggyTask } from "glados/task/defender/piggy";

export class Piggy extends Behavior {
	private _opp: Robot | undefined = undefined;

	protected _stop() {
		this._opp = undefined;
	}

	public check(): Behavior | undefined {
		let role = this._messaging.receiveTrainer(MessageType.roleAssignment);
		return role != undefined && role.name === "Piggy"
			? this
			: undefined;
	}

	protected _updateTask(): TaskAssignment<typeof PiggyTask> {
		let assignment = this._messaging.receiveTrainer(MessageType.roleAssignment);
		if (assignment == undefined || assignment.name !== "Piggy") {
			throw new Error();
		}
		let newOpp = assignment.params[0];
		let restartTask = newOpp !== this._opp;
		this._opp = newOpp;

		debug.set("target", this._opp.id);

		return [PiggyTask, [this._opp], restartTask];
	}
}
